File: lib/models/task_model.dart
import 'package:flutter/foundation.dart';

class Task {
  final String id;
  final String title;
  final DateTime scheduledTime;
  final bool isCompleted;
  final String status; // 'pending', 'ongoing', 'completed'
  final int totalTime; // in seconds for stopwatch

  Task({
    required this.id,
    required this.title,
    required this.scheduledTime,
    this.isCompleted = false,
    this.status = 'pending',
    this.totalTime = 0,
  });

  Task copyWith({
    String? id,
    String? title,
    DateTime? scheduledTime,
    bool? isCompleted,
    String? status,
    int? totalTime,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      scheduledTime: scheduledTime ?? this.scheduledTime,
      isCompleted: isCompleted ?? this.isCompleted,
      status: status ?? this.status,
      totalTime: totalTime ?? this.totalTime,
    );
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      scheduledTime: DateTime.parse(json['scheduledTime']),
      isCompleted: json['isCompleted'] ?? false,
      status: json['status'] ?? (json['isCompleted'] == true ? 'completed' : 'pending'),
      totalTime: json['totalTime'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'scheduledTime': scheduledTime.toIso8601String(),
      'isCompleted': isCompleted,
      'status': status,
      'totalTime': totalTime,
    };
  }
}

File: backend/models/taskModel.js
const mongoose = require('mongoose');

const taskSchema = mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        required: true,
        ref: 'User'
    },
    title: {
        type: String,
        required: [true, 'Please add a task title']
    },
    scheduledTime: {
        type: Date,
        required: true
    },
    isCompleted: {
        type: Boolean,
        default: false
    },
    status: {
        type: String,
        enum: ['pending', 'ongoing', 'completed'],
        default: 'pending'
    },
    totalTime: {
        type: Number, // in seconds
        default: 0
    }
}, {
    timestamps: true
});

module.exports = mongoose.model('Task', taskSchema);
